### DISCUSSION TITLE
Labeled elements in lexer

### DISCUSSION THREAD

**Discussion by MihaMarkic** [antlr/antlr4#4714] [Q&A] [answered]
How does one get access to an element in lexer rule?
Here is an example where I'd like to use only UNQUOTED_STRING part to pass to my method (I'm trying to reference it using 'filePath' name) - together with this token information (I'd like to store token position within file plus the reference path. Looking amateurish at ANTLR4 grammar such construct (filePath=UNQUOTED_STRING should be allowed, but I'm not an expert).

```
REFERENCED_FILEPATH
    : DOUBLE_QUOTE filePath=UNQUOTED_STRING DOUBLE_QUOTE
    {
        AddReferencedFileInfo(_localctx, filePath);
    }
```

TIA

**Comment by jimidle** [comment]
You can’t do this in the lever. It isn’t the right place anyway. Make the
lever as simple as possible, no actions in the parser. Produce the parse
tree, and walk it. You can walk the tree multiple times with different
walkers. Perhaps you need your first walk to build symbols?

Jim

**Reply by MihaMarkic** [reply]
While yes, I could be doing that on parse tree, why not along the way in lexer to avoid double work? Anyway, I think that was a valid approach in at least antlr2 and even antlr4 grammar seemingly support that notation. So I'm curious whether there is even a way to do it like that in antlr4. But anyway, I can use Text and trim double quotes away at worst.
PS. I see that _localCtx is null in such case, instead I'll be using TokenStartLine and TokenStartColumn to figure out the location (I didn't know these two guys exist beforehand).

**Reply by kaby76** [reply, marked as answer]
> .. even antlr4 grammar seemingly support that notation ...

The syntax between lexer and parser rules were identical for Antlr2 and Antlr3 (i.e., a `rule_` referenced an `altList` in [Antlr2](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr2/ANTLRv2Parser.g4#L134C55-L134C62) and [Antlr3](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr3/ANTLRv3Parser.g4#L83C66-L83C73)), but that changed for Antlr4 with [ruleBlock](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr4/ANTLRv4Parser.g4#L217-L219) for parser rules and [lexerRuleBlock](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr4/ANTLRv4Parser.g4#L236-L238) for lexer rules. [element](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr4/ANTLRv4Parser.g4#L301-L306) has a labeled element reference, [lexerElement](https://github.com/antlr/grammars-v4/blob/c82c128d980f4ce46fb3536f87b06b45b9619922/antlr/antlr4/ANTLRv4Parser.g4#L255-L259) doesn't.

**Reply by MihaMarkic** [reply]
Thanks
